File: src/main/java/main/utils/database/mongodb/AbstractMongoDatabase.java
package main.utils.database.mongodb;

import com.mongodb.client.MongoCollection;
import com.mongodb.client.MongoDatabase;
import com.mongodb.client.model.*;
import lombok.Getter;
import main.constants.Database;
import main.utils.database.mongodb.cache.BotBDCache;
import main.utils.database.mongodb.cache.FavouriteTracksCache;
import main.utils.database.mongodb.cache.GuildDBCache;
import main.utils.database.mongodb.databases.GuildDB;
import main.utils.json.GenericJSONField;
import org.bson.BsonArray;
import org.bson.BsonObjectId;
import org.bson.Document;
import org.bson.conversions.Bson;
import org.bson.json.JsonWriterSettings;
import org.bson.types.ObjectId;
import org.json.JSONArray;
import org.json.JSONObject;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.NoSuchElementException;

public abstract class AbstractMongoDatabase {
    private static final Logger logger = LoggerFactory.getLogger(AbstractMongoDatabase.class);

    private final MongoDatabase database;
    @Getter
    private MongoCollection<Document> collection;

    public AbstractMongoDatabase(AbstractMongoDatabase db) {
        this.database = db.database;
        this.collection = db.collection;
    }

    public AbstractMongoDatabase(Database.Mongo db, Database.Mongo collection) {
        final var database = MongoConnectionManager.ins()
                .connect(db).database();
        this.database = database;

        try {
            this.collection = database.getCollection(collection.toString());
        } catch (IllegalArgumentException e) {
            database.createCollection(collection.toString());
            logger.info("Created a new collection with name \"{}\" in {}", collection, db);
            this.collection = database.getCollection(collection.toString());
        }
    }

    public abstract void init();

    public static void initAllCaches() {
        BotBDCache.initCache();
        GuildDBCache.initCache();
        FavouriteTracksCache.initCache();
//        StatisticsDB.INSTANCE.init();
    }

    public static void updateAllCaches() {
//        BotInfoDB.update();
        GuildDB.update();
    }

    MongoCollection<Document> getCollection(String name) {
        return database.getCollection(name);
    }

    protected BsonObjectId addDocument(Document doc) {
        return collection.insertOne(doc).getInsertedId().asObjectId();
    }

    protected void addManyDocuments(List<Document> docs) {
        collection.insertMany(docs);
    }

    protected void upsertManyDocuments(List<Document> docs) {
        List<WriteModel<? extends Document>> bulkWriteModels = new ArrayList<>();
        
        for (var doc : docs) {
            ObjectId id = doc.getObjectId("_id");

            Document oldDoc = null;
            for (var document : collection.find())
                if (document.getObjectId("_id").equals(id))
                    oldDoc = document;

            if (oldDoc != null)
                bulkWriteModels.add(new DeleteOneModel<>(oldDoc));
            bulkWriteModels.add(new InsertOneModel<>(doc));
        }

        collection.bulkWrite(bulkWriteModels);
    }

    protected void upsertManyDocument(List<JSONObject> objects) {
        final List<Document> documents = new ArrayList<>();
        objects.forEach(object -> documents.add(Document.parse(object.toString())));
        upsertManyDocuments(documents);
    }

    protected void upsertDocument(Document doc) {
        ObjectId id = doc.getObjectId("_id");

        Document oldDoc = null;
        for (var document : collection.find())
            if (document.getObjectId("_id").equals(id))
                oldDoc = document;

        if (oldDoc != null)
            removeDocument(oldDoc);
        addDocument(doc);
    }

    protected void upsertDocument(String key, Object value, Document doc) {
        try {
            Document oldDoc = findSpecificDocument(key, value);
            removeDocument(doc);
            addDocument(doc);
        } catch (NoSuchElementException e) {
            addDocument(doc);
        }
    }

    protected void upsertDocument(JSONObject obj) {
        upsertDocument(Document.parse(obj.toString()));
    }

    protected void addDocument(JSONObject obj) {
        collection.insertOne(Document.parse(obj.toString()));
    }

    protected void removeDocument(Document doc) {
        collection.deleteOne(doc);
    }

    protected void removeManyDocuments(Bson bsonQuery) {
        collection.deleteMany(bsonQuery);
    }

    protected void removeManyDocuments(List<Document> docs) {
        List<WriteModel<? extends Document>> bulkWriteModels = new ArrayList<>();
        docs.forEach(doc -> bulkWriteModels.add(new DeleteOneModel<>(doc)));
        collection.bulkWrite(bulkWriteModels);
    }

    /**
     * Remove all documents with the specified key and value
     * @param key The key
     * @param value The value attached to the key
     */
    protected void removeDocument(String key, String value) {
        if (!documentExists(key, value))
            throw new NullPointerException("There is no such document found with key: " + key);

        var iterator = findDocument(key, value);
        while (iterator.hasNext())
            removeDocument(iterator.next());
    }

    protected void removeSpecificDocument(String key, String value) {
        if (!documentExists(key, value))
            throw new NullPointerException("There is no such document found with key: " + key);

        var doc = findDocument(key, value).next();
        removeDocument(doc);
    }

    protected void removeSpecificDocument(String key, Object value) {
        if (!documentExists(key, value))
            throw new NullPointerException("There is no such document found with key: " + key);

        var doc = findDocument(key, value).next();
        removeDocument(doc);
    }

    protected void removeSpecificDocument(GenericJSONField key, Object value) {
        removeSpecificDocument(key.toString(), value);
    }

    protected void updateDocument(Document document, Bson updates) {
        UpdateOptions options = new UpdateOptions().upsert(true);
        collection.updateOne(document, updates, options);
    }

    protected void updateDocument(Document document, List<Bson> updates) {
        UpdateOptions options = new UpdateOptions().upsert(true);
        collection.updateOne(document, updates, options);
    }

    protected <T> void updateDocument(Document document, String key, T newValue) {
        Bson updates = Updates.combine(
                Updates.set(key, newValue),
                Updates.currentTimestamp("lastUpdated")
        );

        updateDocument(document, updates);
    }

    protected void updateDocument(Document document, JSONObject json) {
        final List<Bson> updates = new ArrayList<>();

        for (String key : json.keySet()) {
            if (key.equals("_id")) continue;

            final var value = json.get(key);

            if (value instanceof JSONObject obj) {
                updates.add(Updates.combine(
                        Updates.set(key, Document.parse(obj.toString()))
                ));
                continue;
            } else if (value instanceof JSONArray arr) {
                updates.add(Updates.combine(
                        Updates.set(key, BsonArray.parse(arr.toString()))
                ));
                continue;
            }

            updates.add(Updates.combine(
                    Updates.set(key, json.get(key))
            ));
        }

        updateDocument(document, updates);
    }

    protected <T> void updateDocument(String idName, String idValue, String key, T newValue) {
        Document document = findSpecificDocument(idName, idValue);

        if (document == null)
            throw new NullPointerException("There is no such document with mapping <"+idName+":"+idValue+">");

        updateDocument(document, key, newValue);
    }

    protected boolean documentExists(GenericJSONField key, String value) {
        return findDocument(key.toString(), value).hasNext();
    }

    protected boolean documentExists(GenericJSONField key, Object value) {
        return findDocument(key.toString(), value).hasNext();
    }

    protected boolean specificDocumentExists(GenericJSONField key, Object value) {
        return findSpecificDocument(key.toString(), value) != null;
    }

    protected boolean documentExists(String key, String value) {
        return findDocument(key, value).hasNext();
    }

    protected boolean documentExists(String key, Object value) {
        return findDocument(key, value).hasNext();
    }

    protected boolean documentExists(String key, Document value) {
        return findDocument(key, value).hasNext();
    }

    protected boolean documentExists(String key, JSONObject value) {
        return documentExists(key, Document.parse(value.toString()));
    }

    protected Iterator<Document> findDocument(String key, String value) {
        return collection.find(eq(key, value)).iterator();
    }

    protected Iterator<Document> findDocument(String key, Object value) {
        return collection.find(eq(key, value)).iterator();
    }

    protected Document findSpecificDocument(String key, String value) {
        return collection.find(eq(key, value)).iterator().next();
    }

    protected Document findSpecificDocument(String key, Document value) {
        return collection.find(eq(key, value)).iterator().next();
    }

    protected Document findSpecificDocument(String key, JSONObject value) {
        return collection.find(eq(key, value)).iterator().next();
    }

    protected Document findSpecificDocument(String key, Object value) {
        try {
            return collection.find(eq(key, value)).iterator().next();
        } catch (NoSuchElementException e) {
            return null;
        }
    }

    protected Document findSpecificDocument(GenericJSONField key, Object value) {
        return findSpecificDocument(key.toString(), value);
    }

    protected Iterator<Document> findDocument(String key, Document value) {
        return collection.find(eq(key, value)).iterator();
    }

    protected Iterator<Document> findDocument(String key, JSONObject value) {
        return findDocument(key, Document.parse(value.toString()));
    }

    protected Iterator<Document> findDocument(Document doc) {
        return collection.find(doc).iterator();
    }

    protected Iterator<Document> findDocument(JSONObject obj) {
        return findDocument(Document.parse(obj.toString()));
    }

    protected <T> String getDocument(String key, T value) {
        return getDocument(key, value, false);
    }

    protected <T> String getDocument(String key, T value, boolean indented) {
        Document doc;

        if (value instanceof String s)
            doc = findSpecificDocument(key, s);
        else if (value instanceof Document d)
            doc = findSpecificDocument(key, d);
        else if (value instanceof JSONObject j)
            doc = findSpecificDocument(key, j);
        else if (value instanceof Long l)
            doc = findSpecificDocument(key, l);
        else if (value instanceof Integer i)
            doc = findSpecificDocument(key, i);
        else
            throw new IllegalArgumentException("Invalid value type!");

        if (doc == null)
            throw new NullPointerException("There was no such document with key \""+key+"\" and value \""+value+"\"");

        return documentToJSON(doc, indented);
    }

    protected <T> String getDocuments(String key, T value) {
        return getDocuments(key, value, false);
    }

    protected <T> String getDocuments(String key, T value, boolean indented) {
        Iterator<Document> doc;

        if (value instanceof String s)
            doc = findDocument(key, s);
        else if (value instanceof Document d)
            doc = findDocument(key, d);
        else if (value instanceof JSONObject j)
            doc = findDocument(key, j);
        else
            throw new IllegalArgumentException("Invalid value type!");

        if (doc == null)
            throw new NullPointerException("There was no such document with key \""+key+"\" and value \""+value+"\"");

        StringBuilder sb = new StringBuilder();

        while (doc.hasNext())
            sb.append(documentToJSON(doc.next(), indented)).append("\n");

        return sb.toString();
    }

    protected String documentToJSON(Document doc, boolean indented) {
        return doc.toJson(JsonWriterSettings.builder()
                .indent(indented)
                .build()
        );
    }

    protected String documentToJSON(Document doc) {
        return documentToJSON(doc, false);
    }

    private <TItem> Bson eq(String key, TItem value) {
        return Filters.eq(key, value);
    }
}

File: src/main/kotlin/main/utils/database/mongodb/AbstractMongoDatabase.kt
package main.utils.database.mongodb

import com.mongodb.client.MongoCollection
import com.mongodb.client.MongoDatabase
import com.mongodb.client.model.*
import com.mongodb.client.result.DeleteResult
import dev.minn.jda.ktx.util.SLF4J
import main.constants.database.RobertifyMongoDatabase
import main.utils.database.mongodb.cache.BotDBCache
import main.utils.database.mongodb.cache.FavouriteTracksCache
import main.utils.database.mongodb.cache.redis.guild.GuildRedisCache
import main.utils.json.GenericJSONField
import org.bson.BsonArray
import org.bson.BsonObjectId
import org.bson.Document
import org.bson.conversions.Bson
import org.bson.json.JsonWriterSettings
import org.json.JSONArray
import org.json.JSONObject

abstract class AbstractMongoDatabase {
    companion object {
        /**
         * This value will be used to determine if a document
         * is invalid. If the number of mappings exceed this value
         * more than likely something has gone wrong, and it may lead
         * to a recursive heap overflow. Don't even ask what lead
         * to the circumstances for this implementation
         */
        private const val KEY_MAPPINGS_THRESHOLD = 200
        private val log by SLF4J

        fun initAllCaches() {
            BotDBCache.instance
            FavouriteTracksCache.instance
            GuildRedisCache.initCache()
            updateAllCaches()
        }

        fun updateAllCaches() {

        }
    }

    private val database: MongoDatabase
    var collection: MongoCollection<Document>
        protected set

    constructor(
        db: RobertifyMongoDatabase = RobertifyMongoDatabase.ROBERTIFY_DATABASE,
        collection: RobertifyMongoDatabase
    ) {
        database = MongoConnectionManager.connect(db).database()

        try {
            this.collection = database.getCollection(collection.toString())
        } catch (e: IllegalArgumentException) {
            database.createCollection(collection.toString())
            log.info("Created a new collection with name \"${collection}\" in $db")
            this.collection = database.getCollection(collection.toString())
        }
    }

    constructor(db: AbstractMongoDatabase) {
        database = db.database
        collection = db.collection
    }

    open fun init() {}

    fun addDocument(doc: Document): BsonObjectId =
        collection.insertOne(doc).insertedId?.asObjectId()
            ?: throw NullPointerException("The document could not be inserted due to a null problem!")

    fun addDocument(obj: JSONObject): BsonObjectId = addDocument(Document.parse(obj.toString()))

    fun addManyDocument(docs: Collection<Document>) = collection.insertMany(docs.toMutableList())

    fun upsertManyDocuments(docs: Collection<Document>) {
        val bulkWriteModels = ArrayList<WriteModel<out Document>>()
        docs.forEach { doc ->
            val id = doc.getObjectId("_id")
            var oldDoc: Document? = null

            collection.find().forEach { document ->
                if (document.getObjectId("_id").equals(id))
                    oldDoc = document
            }

            if (oldDoc != null)
                bulkWriteModels.add(DeleteOneModel(oldDoc as Bson))
            bulkWriteModels.add(InsertOneModel(doc))
        }

        collection.bulkWrite(bulkWriteModels)
    }

    fun upsertDocument(oldDoc: Document? = null, doc: Document) {
        when (oldDoc) {
            null -> {
                val id = doc.getObjectId("_id")
                when (val oldDocument = collection.find(Filters.eq("_id", id)).first()) {
                    null -> throw NullPointerException("There is no document to be updated!")
                    else -> {
                        val updatedArr = doc.keys.map { key -> Updates.set(key, doc[key]) }
                        val updates = Updates.combine(updatedArr)
                        collection.updateOne(oldDocument, updates, UpdateOptions().upsert(true))
                    }
                }
            }

            else -> {
                collection.deleteOne(oldDoc)
                collection.insertOne(doc)
            }
        }
    }

    fun upsertDocument(obj: JSONObject) = upsertDocument(doc = Document.parse(obj.toString()))

    protected fun removeDocument(doc: Document) = collection.deleteOne(doc)

    protected fun removeManyDocuments(filter: Bson): DeleteResult {
        return collection.deleteMany(filter)
    }

    protected fun removeDocument(key: String, value: String) = when {
        !documentExists(key, value) -> throw NullPointerException("There is no such document found with key $key")
        else -> {
            val iterator = findDocument(key, value)
            while (iterator.hasNext())
                removeDocument(iterator.next())
        }
    }

    protected fun removeSpecificDocument(key: String, value: Any) = when {
        !documentExists(key, value) -> throw NullPointerException("There is no such document with key: $key")
        else -> removeDocument(findDocument(key, value).next())
    }

    protected fun updateDocument(document: Document, updates: Bson) {
        val options = UpdateOptions().upsert(true)
        collection.updateOne(document, updates, options)
    }

    protected open fun updateDocument(document: Document, updates: List<Bson>) {
        val options = UpdateOptions().upsert(true)
        collection.updateOne(document, updates, options)
    }

    protected open fun <T> updateDocument(document: Document, key: String, newValue: T) {
        val updates = Updates.combine(
            Updates.set(key, newValue),
            Updates.currentTimestamp("lastUpdated")
        )
        updateDocument(document, updates)
    }

    protected open fun updateDocument(document: Document, json: JSONObject) {
        val updates: MutableList<Bson> = java.util.ArrayList()
        for (key in json.keySet()) {
            if (key == "_id") continue
            val value = json[key]
            if (value is JSONObject) {
                updates.add(
                    Updates.combine(
                        Updates.set(key, Document.parse(value.toString()))
                    )
                )
                continue
            } else if (value is JSONArray) {
                updates.add(
                    Updates.combine(
                        Updates.set(key, BsonArray.parse(value.toString()))
                    )
                )
                continue
            }
            updates.add(
                Updates.combine(
                    Updates.set(key, json[key])
                )
            )
        }
        updateDocument(document, updates)
    }

    protected open fun <T> updateDocument(idName: String, idValue: String, key: String, newValue: T) {
        val document = findSpecificDocument(idName, idValue)
            ?: throw NullPointerException("There is no such document with mapping <$idName:$idValue>")
        updateDocument(document, key, newValue)
    }

    protected fun documentExists(key: String, value: String): Boolean =
        findDocument(key, value).hasNext()

    protected fun documentExists(key: String, value: Any): Boolean =
        findDocument(key, value).hasNext()

    private fun validateDocument(document: Document?): Document? {
        if (document == null) return null
        val keys = document.keys
        if (keys.size > KEY_MAPPINGS_THRESHOLD) {
            log.error("A document has exceeded the key mappings threshold!")
            return null
        }
        return document
    }

    protected fun findDocument(key: String, value: String): Iterator<Document> =
        collection.find(Filters.eq(key, value)).iterator()

    protected fun <T> findDocument(key: String, value: T): Iterator<Document> {
        return collection.find(Filters.eq(key, value)).iterator()
    }

    protected fun findSpecificDocument(key: String, value: String): Document? {
        return validateDocument(collection.find(Filters.eq(key, value)).iterator().next())
    }

    protected fun findSpecificDocument(key: String, value: Document): Document? {
        return validateDocument(collection.find(Filters.eq(key, value)).iterator().next())
    }

    protected fun findSpecificDocument(key: String, value: JSONObject): Document? {
        return validateDocument(collection.find(Filters.eq(key, value)).iterator().next())
    }

    fun <T> findSpecificDocument(key: String, value: T): Document? {
        return validateDocument(collection.find(Filters.eq(key, value)).iterator().next())
    }

    protected fun findSpecificDocument(key: GenericJSONField, value: Any): Document? {
        return findSpecificDocument(key.toString(), value)
    }

    protected fun findDocument(key: String, value: Document): Iterator<Document> {
        return collection.find(Filters.eq(key, value)).iterator()
    }

    protected fun findDocument(key: String, value: JSONObject): Iterator<Document> {
        return findDocument(key, Document.parse(value.toString()))
    }

    fun <T> getDocument(key: String, value: T, indented: Boolean = false): String? {
        val doc: Document = when (value) {
            is String -> findSpecificDocument(key, value)
            is Document -> findSpecificDocument(key, value)
            is JSONObject -> findSpecificDocument(key, value)
            is Long -> findSpecificDocument(key, value)
            is Int -> findSpecificDocument(key, value)
            else -> throw IllegalArgumentException("Invalid value type!")
        } ?: return null
        return documentToJSON(doc, indented)
    }

    protected fun <T> getDocuments(key: String, value: T): String {
        return getDocuments(key, value, false)
    }

    protected fun <T> getDocuments(key: String, value: T, indented: Boolean): String {
        val doc: Iterator<Document> = when (value) {
            is String -> findDocument(key, value)
            is Document -> findDocument(key, value)
            is JSONObject -> findDocument(key, value)
            else -> throw IllegalArgumentException("Invalid value type!")
        }
        val sb = StringBuilder()
        while (doc.hasNext()) {
            val nextDoc = doc.next()
            if (validateDocument(nextDoc) != null)
                sb.append(documentToJSON(nextDoc, indented)).append("\n")
        }
        return sb.toString()
    }

    protected fun documentToJSON(doc: Document, indented: Boolean): String {
        return doc.toJson(
            JsonWriterSettings.builder()
                .indent(indented)
                .build()
        )
    }

    fun documentToJSON(doc: Document): String {
        return documentToJSON(doc, false)
    }

    protected fun <TItem> eq(key: String, value: TItem): Bson {
        return Filters.eq(key, value)
    }

    @SafeVarargs
    protected fun <TItem> `in`(key: String, vararg values: TItem): Bson {
        return Filters.`in`(key, *values)
    }

    @SafeVarargs
    protected fun <TItem> notIn(key: String, vararg values: TItem): Bson {
        return Filters.nin(key, *values)
    }
}